src/solver/solverUtils.F90

```diff
@@ -835,114 +852,196 @@ subroutine gridVelocitiesFineLevel_block(useOldCoor, t, sps)
-                ! xx an offset of +1 must be used in the coordinate
-                ! array, because x originally starts at 0 for the
-                ! i, j and k indices.
-
-                do k=1,kke
-                   do j=1,jje
+          ! The original code is elegant but the Tapenade has a difficult time 
+          ! to understand it. Thus, we unfold it and make it easier for the 
+          ! Tapenade.
 
-                      ! Determine the coordinates of the face center,
-                      ! which are stored in xc.
-
-                      xc(1) = fourth*(xx(j+1,k+1,1) + xx(j,k+1,1) &
-                           +         xx(j+1,k,  1) + xx(j,k,  1))
-                      xc(2) = fourth*(xx(j+1,k+1,2) + xx(j,k+1,2) &
-                           +         xx(j+1,k,  2) + xx(j,k,  2))
-                      xc(3) = fourth*(xx(j+1,k+1,3) + xx(j,k+1,3) &
-                           +         xx(j+1,k,  3) + xx(j,k,  3))
-
-                      ! Determine the coordinates relative to the
-                      ! center of rotation.
-
-                      xxc(1) = xc(1) - rotCenter(1)
-                      xxc(2) = xc(2) - rotCenter(2)
-                      xxc(3) = xc(3) - rotCenter(3)
-
-                      ! Determine the rotation speed of the face center,
-                      ! which is omega*r.
-
-                      sc(1) = rotRate(2)*xxc(3) - rotRate(3)*xxc(2)
-                      sc(2) = rotRate(3)*xxc(1) - rotRate(1)*xxc(3)
-                      sc(3) = rotRate(1)*xxc(2) - rotRate(2)*xxc(1)
-
-                      ! Determine the coordinates relative to the
-                      ! rigid body rotation point.
-
-                      xxc(1) = xc(1) - rotationPoint(1)
-                      xxc(2) = xc(2) - rotationPoint(2)
-                      xxc(3) = xc(3) - rotationPoint(3)
-
-                      ! Determine the total velocity of the cell face.
-                      ! This is a combination of rotation speed of this
-                      ! block and the entire rigid body rotation.
-
-                      sc(1) = sc(1) + velxGrid           &
-                           + derivRotationMatrix(1,1)*xxc(1) &
-                           + derivRotationMatrix(1,2)*xxc(2) &
-                           + derivRotationMatrix(1,3)*xxc(3)
-                      sc(2) = sc(2) + velyGrid           &
-                           + derivRotationMatrix(2,1)*xxc(1) &
-                           + derivRotationMatrix(2,2)*xxc(2) &
-                           + derivRotationMatrix(2,3)*xxc(3)
-                      sc(3) = sc(3) + velzGrid           &
-                           + derivRotationMatrix(3,1)*xxc(1) &
-                           + derivRotationMatrix(3,2)*xxc(2) &
-                           + derivRotationMatrix(3,3)*xxc(3)
-
-                      ! Store the dot product of grid velocity sc and
-                      ! the normal ss in sFace.
+          ! i-direction
```

Comment:
This is a bigger chunk of "unravelled" pointers loop, please take a look to see if this is good to merge

Reply:
There is no way to track the coverage in Fortran, but `gridVelocitiesFineLevel_block` is called by `gridVelocitiesFineLevel` and that subroutine is used by both the rotating frame and the unsteady cases, so if anything is broken here we should have major failures or errors in the tests we added.
Any opinions on this @SichengHe ?